### PR TITLE
TandemFoil: T_max=3 — ultra-ultra-short cosine cycles

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

T_max controls the period of cosine annealing cycles. The scheduler runs per-batch, and with ~750 steps/epoch:
- T_max=10 → ~75 complete cosine cycles/epoch (current best: 75.59)
- T_max=5 → ~150 complete cosine cycles/epoch (being tested separately)
- T_max=3 → ~250 complete cosine cycles/epoch (this experiment)

T_max=10 already beat T_max=20, T_max=15, and T_max=30, establishing that more rapid LR oscillation is better for this dataset. The question is: does this trend continue to T_max=3 (250 restarts/epoch), or is there a floor where LR oscillation becomes too rapid and counterproductive?

At T_max=3, the LR oscillates 250 complete cycles per epoch. This is an extremely aggressive schedule that functions almost like a high-frequency noise injection on the learning rate. If the trend holds, this could yield a new best. If not, it tells us the optimal is somewhere between T_max=3 and T_max=10.

## Instructions

Run exactly the following command with `--cosine_t_max 3`:

```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion \
  --lr 3e-4 \
  --cosine_t_max 3 \
  --no-use-ema \
  --model_slices 64 \
  --enable-fourier \
  --enable-te-coord-frame \
  --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 \
  --wandb_group tandem-tmax-sweep
```

No code changes needed — this is a hyperparameter sweep using the existing baseline config.

## Baseline

- **Primary metric:** `val_primary/surface_pressure_mae` (lower is better)
- **Current best:** 75.59 (val) / 72.12 (test)
- **Best PR:** #2490 (frieren — T_max=10, Fourier + physics + no-EMA, slices=64, Lion lr=3e-4, 14 epochs)
- **W&B run:** 77yoba65
- **T_max sweep context:** T_max=10 (75.59) > T_max=20 (77.00) > T_max=15 (80.23). More rapid = better so far.
- **Reproduce baseline:**
  ```bash
  cd target/icml2026 && python train.py \
    --dataset tandemfoil \
    --optimizer lion \
    --lr 3e-4 \
    --cosine_t_max 10 \
    --no-use-ema \
    --model_slices 64 \
    --enable-fourier \
    --enable-te-coord-frame \
    --enable-cp-panel \
    --enable-cp-panel-tandem-only \
    --asinh-pressure \
    --residual-prediction \
    --enable-pressure-prior-addition \
    --epochs 999
  ```

## Key Constraints
- `--epochs 999` is MANDATORY
- 4L/256d is too slow for TandemFoil — stick with 3L/192d (default)
- All physics features above are synergistic with Fourier — keep them all enabled
- Target to beat: **75.59 val_primary/surface_pressure_mae**

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing — T_max=3 at 3L/192d is obsolete. Baseline moved to 5L/256d (52.81). mikasa, please pick up your next assignment.

**Comment by tcapelle** [comment]
STUDENT mikasa:

## Results

### Primary Metrics

| Metric | T_max=3 (this run) | T_max=10 baseline (PR #2490) | Change |
|--------|-------------------|------------------------------|--------|
| **val_primary/surface_pressure_mae** | **47.86** (ep108) | 75.59 (ep14) | **-36.7%** |
| **test_primary/surface_pressure_mae** | **51.14** | 72.12 | **-29.1%** |

### Per-Split Test MAE

| Split | T_max=3 | T_max=10 baseline | Change |
|-------|---------|-------------------|--------|
| single_in_dist | 58.03 | 72.33 | -19.8% |
| geom_camber_rc | 55.99 | 76.01 | -26.3% |
| geom_camber_cruise | 47.44 | 70.80 | -33.0% |
| re_rand | 43.11 | 69.34 | -37.8% |

### Training Details

- **Epochs:** 120 (180-min timeout)
- **W&B run:** `ekxsfwgl`
- **Peak VRAM:** ~38 GB (GPU 0)
- **Command:**
  ```bash
  cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
    --dataset tandemfoil --optimizer lion --lr 3e-4 \
    --cosine-t-max 3 --no-use-ema --model-slices 64 \
    --enable-fourier --enable-te-coord-frame --enable-cp-panel \
    --enable-cp-panel-tandem-only --asinh-pressure \
    --residual-prediction --enable-pressure-prior-addition \
    --epochs 999 --agent mikasa \
    --wandb-name "mikasa/tandem-tmax3" \
    --wandb-group tandem-tmax-sweep
  ```

### What happened

**T_max=3 is a clear winner**, achieving the best TandemFoil results by a wide margin. The hypothesis that "more rapid LR oscillation is better" continues to hold even at the extreme of ~250 cosine cycles per epoch.

Key observations:
1. **Massive improvement over baseline:** 36.7% val improvement, 29.1% test improvement over T_max=10.
2. **All splits improved:** Every test split saw 20-38% improvement, with the largest gains on re_rand (-37.8%) and geom_camber_cruise (-33.0%).
3. **High variance characteristic:** T_max=3 oscillates wildly (range ~48-95 val MAE) compared to T_max=10's smoother curve. The "effective schedule" at T_max=3 acts more like high-frequency noise injection on a near-constant LR, which appears beneficial for escaping sharp minima.
4. **Descending trough pattern:** Best values formed a clear descending sequence: 68.46 (ep17) → 56.62 (ep27) → 55.82 (ep52) → 53.28 (ep56) → 51.18 (ep60) → 50.87 (ep65) → 48.22 (ep80) → 47.86 (ep108). Each oscillation cycle found a deeper minimum.
5. **Extended training was essential:** The baseline only had 14 epochs. T_max=3 at epoch 14 was 74.86 — only marginally better than T_max=10's 75.59. The dramatic gap opened up with extended training (120 epochs), where T_max=3's noise-injection effect accumulated advantage.
6. **Still improving at cutoff:** The trough at ep108 (47.86) was still the best, suggesting more training could push lower.

### Suggested follow-ups

1. **T_max=5 comparison:** T_max=5 is being tested separately — the three-way comparison (T_max=3/5/10) would map the optimal oscillation frequency.
2. **T_mult=2 growing cycles (from T_0=10):** The SGDR paper's strongest result used growing cycle lengths. Starting with T_0=10 and doubling each restart (T_mult=2) is an unexplored direction that could combine the fast early exploration of short cycles with the stable convergence of longer later cycles.
3. **Even longer training:** T_max=3 was clearly still improving at 120 epochs. A 360-min budget run could push significantly further.
4. **T_max=1 extreme:** If T_max=3 beats T_max=10, does T_max=1 beat T_max=3? At T_max=1, the LR effectively becomes constant at the midpoint — worth testing to find the actual floor.
